server/src/main/scala/org/http4s/server/middleware/FormDataDecoder.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s
+package server
+package middleware
+
+import cats.Applicative
+import cats.data.Validated.Valid
+import cats.data.{Chain, ValidatedNel}
+import cats.effect.Sync
+import cats.implicits._
+
+/**
+  * A simple middle ware that uses [[QueryParamDecoder]] to decode values in [[org.http4s.UrlForm]]
+  *
+  * @example
+  * {{{
+  * scala> import cats.implicits._
+  * scala> import cats.data._
+  * scala> import org.http4s.server.middleware.FormDataDecoder._
+  * scala> import org.http4s.ParseFailure
+  * scala> case class Foo(a: String, b: Boolean)
+  * scala> case class Bar(fs: List[Foo], f: Foo, d: Boolean)
+  * scala>
+  * scala> implicit val fooMapper = (
+  *      |   field[String]("a"),
+  *      |   field[Boolean]("b")
+  *      | ).mapN(Foo.apply)
+  * scala>
+  * scala> val barMapper = (
+  *      |   list[Foo]("fs"),
+  *      |   nested[Foo]("f"),
+  *      |   field[Boolean]("d")
+  *      | ).mapN(Bar.apply)
+  * scala>
+  * scala> barMapper(
+  *      |   Map(
+  *      |    "fs[].a" -> Chain("a1", "a2"),
+  *      |    "fs[].b" -> Chain("true", "false"),
+  *      |    "f.a" -> Chain("fa"),
+  *      |    "f.b" -> Chain("false"),
+  *      |    "d" -> Chain("true"))
+  *      | )
+  * res1: ValidatedNel[ParseFailure, Bar] = Valid(Bar(List(Foo(a1,true), Foo(a2,false)),Foo(fa,false),true))
+  * }}}
+  */
+sealed trait FormDataDecoder[A] {
+  def apply(data: Map[String, Chain[String]]): ValidatedNel[ParseFailure, A]
+
+  def mapValidated[B](f: A => ValidatedNel[ParseFailure, B]): FormDataDecoder[B] =
+    FormDataDecoder(this(_).andThen(f))
+
+}
+
+object FormDataDecoder {
+  type FormData = Map[String, Chain[String]]
```

Comment:
I chose these simply because `UrlForm` uses this data type, and the purpose of this class is to be chained after `UrlForm` so that users can easily write:
```scala
    HttpRoutes
     .of[F] {
       case req @ POST -> Root =>
         req.as[MyEntity].flatMap { entity =>
           Ok()
         }
     }
 ```

Reply:
Just to be clear. I am open to change it. I can also change it in both UrlForm and here. Let me know.

Reply:
I think we can't change `UrlForm` in 0.21 because of bincompat, but we should consider it for 1.0.

The question then is whether we should make this like `UrlForm` in 0.21, or make it look like the future.  Curious what other contributors think on this.

Reply:
I think consistency within the release is preferable; if users of UrlForm need to change for 1.0, then changing code using *this* in the same manner shouldn't be a burden